### PR TITLE
Sketch out new credential / authentication pure module.

### PR DESCRIPTION
**Pull request by davidsunglee**
#### What type of PR is this?
Enhancement.

#### What does this PR do / why is it needed ?
Sketch out a top-level module for credentials and authentication to be shared across stores including relational and service store.

#### Which issue(s) this PR fixes:
Issue TBD.

#### Other notes for reviewers:
None

#### Does this PR introduce a user-facing change?
No

### CODE REVIEW

**Comment by davidsunglee** [comment]
Model

File: legend-engine-authentication-pure/src/main/resources/core_authentication/authentication/metamodel.pure [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+{
+  discoveryUrl: String[1];
+}
+
+Class meta::pure::credential::OAuthTokenProvider extends CredentialProvider
+{
+  url: String[1];
+}
+
+Class <<typemodifiers.abstract>> meta::pure::authentication::Authentication
```
Comment (epsstan) [inline legend-engine-authentication-pure/src/main/resources/core_authentication/authentication/metamodel.pure:54]: ```Authentication``` as a generic concept makes sense.  It's how a platform component (stores, persistence, mastery etc) can describe authentication behavior.

In some cases authentication is disabled. For e.g a database used for testing. We can either support this by making ```Authentication``` an optional attribute of ```Connection```

File: legend-engine-authentication-pure/src/main/resources/core_authentication/authentication/metamodel.pure [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+{
+}
+
+Class meta::pure::authentication::UsernamePasswordAuthentication extends Authentication
```
Comment (epsstan) [inline legend-engine-authentication-pure/src/main/resources/core_authentication/authentication/metamodel.pure:58]: ```Authentication``` types have 2 classes of attributes.  
* Attributes such as username etc that are not sensitive and can be in-lined in the model 
* Attributes such as credentials which are sensitive. These cannot be in-lined in the model and therefore we have a ```Credential``` type that describes where and how these values can be obtained

File: legend-engine-authentication-pure/src/main/resources/core_authentication/authentication/metamodel.pure [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+{
+  username: String[1];
+  password: PlainTextCredential[0..1];
```
Comment (epsstan) [inline legend-engine-authentication-pure/src/main/resources/core_authentication/authentication/metamodel.pure:61]: Should we make password a required property ?

File: legend-engine-authentication-pure/src/main/resources/core_authentication/authentication/metamodel.pure [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+Class meta::pure::credential::JsonWebTokenCredential extends Credential
+{
+}
+
+Class <<typemodifiers.abstract>> meta::pure::credential::CredentialProvider
+{
+}
+
+Class meta::pure::credential::VaultCredentialProvider extends CredentialProvider
```
Comment (epsstan) [inline legend-engine-authentication-pure/src/main/resources/core_authentication/authentication/metamodel.pure:39]: Will we have subtypes of ```VaultCredentialProvider``` ?

```
AWSVaultCredentialProvider extends VaultCredentialProvider
{
   name: String[1];
   arn : String[1];
   version: String[1];
}
```
```
EnvironmentVaultCredentialProvider extends VaultCredentialProvider
{
   name: String[1];
}
```

**Comment by davidsunglee** [comment]
Screen Shot 2022-11-08 at 12 53 04 PM

**Comment by finos-admin** [comment]
This PR is stale because it has been open for 30 days with no activity. Please remove stale label or add any comment to keep this open. Otherwise this will be closed in 5 days.

**Comment by finos-admin** [comment]
This PR was closed because it has been inactive for 35 days. Please re-open if this PR is still relevant.

**Pull request closed without merging**